### PR TITLE
[Feature][Connector-V2] Iceberg-sink supports writing data to branches

### LINKED ISSUE #6686: [Feature][Connector-V2] Iceberg-sink supports writing data to branches
### Search before asking

- [X] I had searched in the [feature](https://github.com/apache/seatunnel/issues?q=is%3Aissue+label%3A%22Feature%22) and found no similar feature requirement.

### Description

Iceberg-sink supports writing data to branches

### Usage Scenario

_No response_

### Related issues

_No response_

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by sunxiaojian**
### Purpose of this pull request

closed https://github.com/apache/seatunnel/issues/6686

### Check list

* [ ] If any new Jar binary package adding in your PR, please add License Notice according
  [New License Guide](https://github.com/apache/seatunnel/blob/dev/docs/en/contribution/new-license.md)
* [ ] If necessary, please update the documentation to describe the new feature. https://github.com/apache/seatunnel/tree/dev/docs
* [ ] If you are contributing the connector code, please check that the following files are updated:
  1. Update change log that in connector document. For more details you can refer to [connector-v2](https://github.com/apache/seatunnel/tree/dev/docs/en/connector-v2)
  2. Update [plugin-mapping.properties](https://github.com/apache/seatunnel/blob/dev/plugin-mapping.properties) and add new connector information in it
  3. Update the pom file of [seatunnel-dist](https://github.com/apache/seatunnel/blob/dev/seatunnel-dist/pom.xml)
* [ ] Update the [`release-note`](https://github.com/apache/seatunnel/blob/dev/release-note.md).

### CODE REVIEW

File: seatunnel-e2e/seatunnel-connector-v2-e2e/connector-iceberg-e2e/src/test/java/org/apache/seatunnel/e2e/connector/iceberg/IcebergSinkWithBranchIT.java [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+                    extractFiles();
+                }
+
+                private void extractFiles() {
+                    ProcessBuilder processBuilder = new ProcessBuilder();
+                    processBuilder.command(
+                            "sh", "-c", "cd " + CATALOG_DIR + " && tar -zxvf " + NAMESPACE_TAR);
+                    try {
+                        Process process = processBuilder.start();
+                        // 等待命令执行完成
```
Comment (TaoZex) [inline seatunnel-e2e/seatunnel-connector-v2-e2e/connector-iceberg-e2e/src/test/java/org/apache/seatunnel/e2e/connector/iceberg/IcebergSinkWithBranchIT.java]: please use English comment

File: seatunnel-e2e/seatunnel-connector-v2-e2e/connector-iceberg-e2e/src/test/java/org/apache/seatunnel/e2e/connector/iceberg/IcebergSinkWithBranchIT.java [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+        tableLoader.open();
+        // from branch
+        Table table = tableLoader.loadTable();
+        TableScan branchRead = table.newScan().useRef("commit-branch");
+        CloseableIterable<FileScanTask> fileScanTasks = branchRead.planFiles();
+        fileScanTasks.forEach(
+                fileScanTask -> {
+                    try {
+                        DataFile file = fileScanTask.file();
+                        // 构建文件的输入流
```
Comment (TaoZex) [inline seatunnel-e2e/seatunnel-connector-v2-e2e/connector-iceberg-e2e/src/test/java/org/apache/seatunnel/e2e/connector/iceberg/IcebergSinkWithBranchIT.java]: same as above

**Comment by sunxiaojian** [comment]
@hailin0 PTAL

File: seatunnel-e2e/seatunnel-connector-v2-e2e/connector-iceberg-e2e/src/test/java/org/apache/seatunnel/e2e/connector/iceberg/IcebergSinkWithBranchIT.java (hunk 2 above)
Comment (sunxiaojian) [reply seatunnel-e2e/seatunnel-connector-v2-e2e/connector-iceberg-e2e/src/test/java/org/apache/seatunnel/e2e/connector/iceberg/IcebergSinkWithBranchIT.java]: done

**Review by hailin0 — COMMENTED** [review]
please update docs

File: seatunnel-e2e/seatunnel-connector-v2-e2e/connector-iceberg-e2e/src/test/java/org/apache/seatunnel/e2e/connector/iceberg/IcebergSinkWithBranchIT.java [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+                            log.error("Extract files failed with exit code " + exitCode);
+                        }
+                    } catch (IOException | InterruptedException e) {
+                        e.printStackTrace();
+                    }
+                }
+            };
+
+    @TestTemplate
+    public void testInsertAndCheckDataE2e(TestContainer container)
```
Comment (hailin0) [inline seatunnel-e2e/seatunnel-connector-v2-e2e/connector-iceberg-e2e/src/test/java/org/apache/seatunnel/e2e/connector/iceberg/IcebergSinkWithBranchIT.java:133]: How to validate `iceberg.table.commit-branch="commit-branch"`

Comment (sunxiaojian) [reply seatunnel-e2e/seatunnel-connector-v2-e2e/connector-iceberg-e2e/src/test/java/org/apache/seatunnel/e2e/connector/iceberg/IcebergSinkWithBranchIT.java:133]: > How to validate `iceberg.table.commit-branch="commit-branch"`

'loadDataFromIcebergTableBranch()' uses 'table.newScan().useRef(commitBranch)' to specify that data should be read from a branch.

**Comment by sunxiaojian** [comment]
@hailin0 PTAL

**Comment by hailin0** [comment]
please check ci error or retry

**Comment by sunxiaojian** [comment]
> please check ci error or retry

here should be a problem with this e2e, it has never been successful in trying again

image

**Comment by sunxiaojian** [comment]
> please check ci error or retry

@hailin0 done

File: seatunnel-connectors-v2/connector-iceberg/src/main/java/org/apache/seatunnel/connectors/seatunnel/iceberg/config/SinkConfig.java [hunk 4]
```diff
@@ -109,13 +109,20 @@ public class SinkConfig extends CommonConfig {
                     .defaultValue(DataSaveMode.APPEND_DATA)
                     .withDescription("data save mode");
 
+    public static final Option<String> TABLES_DEFAULT_COMMIT_BRANCH =
```
Comment (hailin0) [inline seatunnel-connectors-v2/connector-iceberg/src/main/java/org/apache/seatunnel/connectors/seatunnel/iceberg/config/SinkConfig.java:112]: add to this file?
https://github.com/apache/seatunnel/blob/dev/seatunnel-connectors-v2/connector-iceberg/src/main/java/org/apache/seatunnel/connectors/seatunnel/iceberg/sink/IcebergSinkFactory.java#L57

Comment (sunxiaojian) [reply seatunnel-connectors-v2/connector-iceberg/src/main/java/org/apache/seatunnel/connectors/seatunnel/iceberg/config/SinkConfig.java:112]: > add to this file? https://github.com/apache/seatunnel/blob/dev/seatunnel-connectors-v2/connector-iceberg/src/main/java/org/apache/seatunnel/connectors/seatunnel/iceberg/sink/IcebergSinkFactory.java#L57

@hailin0 done

File: seatunnel-e2e/seatunnel-connector-v2-e2e/connector-iceberg-e2e/src/test/java/org/apache/seatunnel/e2e/connector/iceberg/IcebergSinkWithBranchIT.java [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+                                        new Path(file.path().toString()), new Configuration());
+                        try (ParquetReader<Object> reader =
+                                AvroParquetReader.builder(inputFile).build()) {
+                            Object record;
+                            while ((record = reader.read()) != null) {
+                                results.add(record);
+                            }
+                        }
+                    } catch (IOException e) {
+                        e.printStackTrace();
```
Comment (hailin0) [inline seatunnel-e2e/seatunnel-connector-v2-e2e/connector-iceberg-e2e/src/test/java/org/apache/seatunnel/e2e/connector/iceberg/IcebergSinkWithBranchIT.java]: add to log message

File: seatunnel-e2e/seatunnel-connector-v2-e2e/connector-iceberg-e2e/src/test/java/org/apache/seatunnel/e2e/connector/iceberg/IcebergSinkWithBranchIT.java [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+                    try {
+                        Process process = processBuilder.start();
+                        int exitCode = process.waitFor();
+                        if (exitCode == 0) {
+                            log.info("Extract files successful.");
+                        } else {
+                            log.error("Extract files failed with exit code " + exitCode);
+                        }
+                    } catch (IOException | InterruptedException e) {
+                        e.printStackTrace();
```
Comment (hailin0) [inline seatunnel-e2e/seatunnel-connector-v2-e2e/connector-iceberg-e2e/src/test/java/org/apache/seatunnel/e2e/connector/iceberg/IcebergSinkWithBranchIT.java]: add to log message

**Review by hailin0 — COMMENTED** [review]
LGTM

File: seatunnel-e2e/seatunnel-connector-v2-e2e/connector-iceberg-e2e/src/test/java/org/apache/seatunnel/e2e/connector/iceberg/IcebergSinkWithBranchIT.java (hunk 6 above)
Comment (sunxiaojian) [reply seatunnel-e2e/seatunnel-connector-v2-e2e/connector-iceberg-e2e/src/test/java/org/apache/seatunnel/e2e/connector/iceberg/IcebergSinkWithBranchIT.java]: > add to log message

@hailin0 done

**Review by TaoZex — APPROVED** [review]
+1

**Review by hailin0 — APPROVED** [review]

**Review by EricJoy2048 — APPROVED** [review]

**Pull request merged**